segpy/field_types.py

```diff
@@ -1,29 +1,86 @@
-class Int16(int):
-
-    MINIMUM = -32768
-    MAXIMUM = 32767
-    SIZE = 2
-    SEG_Y_TYPE = 'int16'
-
-    def __new__(cls, *args, **kwargs):
-        instance = super().__new__(cls, *args, **kwargs)
-        if not (Int16.MINIMUM <= instance <= Int16.MAXIMUM):
-            raise ValueError("{} value {!r} outside range {} to {}".format(cls.__name__, instance,
-                                                                        cls.MINIMUM, cls.MAXIMUM))
+from .datatypes import LIMITS, SEG_Y_TYPE_TO_CTYPE, size_in_bytes
+
+
+class IntFieldMeta(type):
+    """Metaclass for signed and unsigned int fields.
+    """
+
+    def class_new(cls, *args, **kwargs):
+        instance = super(cls, cls).__new__(cls, *args, **kwargs)
+        if not (cls.MINIMUM <= instance <= cls.MAXIMUM):
+            raise ValueError("{} value {!r} outside range {} to {}".format(
+                cls.__name__, instance,
+                cls.MINIMUM, cls.MAXIMUM))
         return instance
 
+    def __new__(cls, name, bases, namespace,
+                seg_y_type,
+                min_value=None,
+                max_value=None,
+                **kwargs):
+
+        bases = bases + (int,)
+
+        if min_value is None:
+            min_value = LIMITS[seg_y_type].min
+
+        if max_value is None:
+            max_value = LIMITS[seg_y_type].max
+
+        namespace.update({
+            'SEG_Y_TYPE': seg_y_type,
+            'MINIMUM': min_value,
+            'MAXIMUM': max_value,
+            'SIZE': size_in_bytes(SEG_Y_TYPE_TO_CTYPE[seg_y_type]),
+            '__new__': cls.class_new,
+        })
+        return super().__new__(cls, name, bases, namespace, **kwargs)
+
+    def __init__(cls, name, bases, namespace, *args, **kwargs):
+        super().__init__(name, bases, namespace)
+
+
+class Int16(metaclass=IntFieldMeta,
+            seg_y_type='int16'):
+    pass
 
-class Int32(int):
 
-    MINIMUM = -2147483648
-    MAXIMUM = 2147483647
-    SIZE = 4
-    SEG_Y_TYPE = 'int32'
+class UInt16(metaclass=IntFieldMeta,
+             seg_y_type='uint16'):
+    pass
 
-    def __new__(cls, *args, **kwargs):
-        instance = super().__new__(cls, *args, **kwargs)
-        if not (Int32.MINIMUM <= instance <= Int32.MAXIMUM):
-            raise ValueError("{} value {!r} outside range {} to {}".format(cls.__name__, instance,
-                                                                        cls.MINIMUM, cls.MAXIMUM))
+
+class Int32(metaclass=IntFieldMeta,
+            seg_y_type='int32'):
+    pass
+
+
+class UInt32(metaclass=IntFieldMeta,
+             seg_y_type='uint32'):
+    pass
+
+
+class IntEnumFieldMeta(IntFieldMeta):
```

Comment:
Should we be leveraging `IntEnum` from the standard library here in some way? https://docs.python.org/3/library/enum.html#intenum

The `Enum` class is specifically designed to support multiple-inheritance: https://docs.python.org/3/library/enum.html#others

Reply:
We probably could. I started off in that direction, but it entailed a lot of extra code (i.e. new enums for all relevant fields) that seemed superfluous for my purposes. Could we leave this as a future exercise?

Reply:
Actually, on thinking about this more, I may have been approaching it from the wrong direction. Let me see what I can put together.

Reply:
Enum type per field is probably appropriate. It's more code, but it all carries value and allows us to easily print textual equivalents of numeric values in diagnostics and such like.

Reply:
There seem to be two high-level approaches we could take here. The first  - similar to what you hinted at in the discussion of `DATA_SAMPLE_FORMAT_TO_SEG_Y_TYPE` - is that we could define the `IntEnum` and the field-type as the same class, e.g.:
```
class DataSampleFormat(metaclass=IntEnumFieldMeta):
    IBM = 1
    Int32 = 2
    . . .
```
In this case `DataSampleFormat` would be a subclass of `IntEnum` and it would have the extra class-level features of a field (min/max value, etc.). It could be used as field in a header.

In order to avoid circular dependencies, this approach would force us the consider where `DataSampleFormat` should be defined. Right now it's defined in `datatypes.py` which is imported by `field_types.py`...this is where `IntEnumFieldMeta` would (presumably) be defined.

The alternative is that we define the enum and field separately:
```
class DataSampleFormat(enum.IntEnum):
    IBM = 1
    . . . 

class DataSampleFormatField(metaclass=IntEnumFieldMeta,
                                                  enum=DataSampleFormat):
     pass
```

This feels like a cleaner separation of concerns: the enum on the one hand, the field on the other. But it's also more verbose with two constructs per logical field rather than one. It's more flexible, at least in principle, because you could define another field with the same enum using e.g. a different storage size. If SegY really is that hairy in the wild, this could be useful.

Reply:
Hmmm. Creating a field-type that is also an `IntEnum` subclass is tricky, and might actually be impossible. Our fields have class-level attributes for things like min/max value, but `Enums` determine their constituents by looking for class-level attributes! If we try to add the class-level attributes after the enum-scanning machinery runs, we get an exception:
```
File "python3.6/enum.py", line 436, in _get_mixins_
    raise TypeError("Cannot extend enumerations")
```